### DISCUSSION TITLE
How to add content to already loaded webpage?

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3848] [Q&A]
Hi! How to add content to already loaded webpage? I`m trying smth like this:

```
Browser.LoadHtml(Browser.GetSourceAsync().Result +
             "<extra_data>some new content</extra_data>", Browser.Address);

string checkWhatWeHaveFinally = Browser.GetSourceAsync().Result;
```

But not getting the entire code (with added data) as a result.

**Comment by amaitland** [comment]
Adding new content with JavaScript is the approach I'd usually go with.

The example you've provided you aren't waiting for the page to load so you probably get the original source.

**Reply by assopri** [reply]
1. How I can wait for page to be loaded after LoadHtml(...) call?
2. Any examples how to add new content with JS? At least - what is the general approach (adding to some last element in dom structure or...)? Is waiting for page to be loaded needed in this case?

**Reply by mitchcapper** [reply]
there is a loadurl async call you can await for the navigation to occur.     As for adding new content with JS the simple way would be to encode your HTML into a base 64 string in c# then call the javascript function evaluation and in javascript you just use atob to decode it.   Technically cefsharp supports passing more complex data structures but this may be an easy and simple way to start.  You could then insert the HTML into the DOM with standard javascript.   See the getting started faq for making the JS calls.

**Reply by assopri** [reply]
> there is a loadurl async call you can await for the navigation to occur. As for adding new content with JS the simple way would be to encode your HTML into a base 64 string in c# then call the javascript function evaluation and in javascript you just use atob to decode it. Technically cefsharp supports passing more complex data structures but this may be an easy and simple way to start. You could then insert the HTML into the DOM with standard javascript. See the getting started faq for making the JS calls.

Hello! Thank you for reply. As for loadurl isn`t it mutually exclusive alternative to LoadHtml() ? I`m just messed up with their interaction in this case. I think, if I use LoadUrl - it will reset the pagecode, which I was setting with LoadHtml, isnt it?

**Reply by amaitland** [reply]
Any of the `Load` methods will load a new page. If you need to append html to the existing page then executing javascript is an easy option.
